pkg/specgen/specgen.go

```diff
@@ -0,0 +1,363 @@
+	// Optional.
+	StaticIPv6 *net.IP
+	// StaticMAC is a static MAC address to set in the container.
+	// Only available if ConfigureNetNS is true.
+	// Optional.
+	StaticMAC *net.HardwareAddr
+	// PortBindings is a set of ports to map into the container.
+	// Only available if ConfigureNetNS is true.
+	// Optional.
+	PortMappings []ocicni.PortMapping
+	// PublishImagePorts will publish ports specified in the image to random
+	// ports outside.
+	// Requires Image to be set.
+	PublishImagePorts bool
+	// CNINetworks is a list of CNI networks to join the container to.
+	// If this list is empty, the default CNI network will be joined
+	// instead. If at least one entry is present, we will not join the
+	// default network (unless it is part of this list).
+	// Only available if ConfigureNetNS is true.
+	// Optional.
+	CNINetworks []string
+	// UseImageResolvConf indicates that resolv.conf should not be managed
+	// by Podman, but instead sourced from the image.
+	// Conflicts with DNSServer, DNSSearch, DNSOption.
+	UseImageResolvConf bool
+	// DNSServer is a set of DNS servers that will be used in the
+	// container's resolv.conf, replacing the host's DNS Servers which are
+	// used by default.
+	// Conflicts with UseImageResolvConf.
+	// Optional.
+	DNSServer []net.IP
+	// DNSSearch is a set of DNS search domains that will be used in the
+	// container's resolv.conf, replacing the host's DNS search domains
+	// which are used by default.
+	// Conflicts with UseImageResolvConf.
+	// Optional.
+	DNSSearch []string
+	// DNSOption is a set of DNS options that will be used in the
+	// container's resolv.conf, replacing the host's DNS options which are
+	// used by default.
+	// Conflicts with UseImageResolvConf.
+	// Optional.
+	DNSOption []string
+	// UseImageHosts indicates that /etc/hosts should not be managed by
+	// Podman, and instead sourced from the image.
+	// Conflicts with HostAdd.
+	UseImageHosts bool
+	// HostAdd is a set of hosts which will be added to the container's
+	// /etc/hosts file.
+	// Conflicts with UseImageHosts.
+	// Optional.
+	HostAdd []string
+}
+
+// ContainerResourceConfig contains information on container resource limits.
+type ContainerResourceConfig struct {
+	// ResourceLimits are resource limits to apply to the container.
+	// Can only be set as root on cgroups v1 systems, but can be set as
+	// rootless as well for cgroups v2.
+	// Optional.
+	ResourceLimits *spec.LinuxResources
+	// Rlimits are POSIX rlimits to apply to the container.
+	// Optional.
+	Rlimits []spec.POSIXRlimit
+	// OOMScoreAdj adjusts the score used by the OOM killer to determine
+	// processes to kill for the container's process.
+	// Optional.
+	OOMScoreAdj *int
+}
+
+// SpecGenerator creates an OCI spec and Libpod configuration options to create
+// a container based on the given configuration.
+type SpecGenerator struct {
+	ContainerBasicConfig
+	ContainerStorageConfig
+	ContainerSecurityConfig
+	ContainerCgroupConfig
+	ContainerNetworkConfig
```

Comment:
it's also similarly odd that cgroup and net namespaces get their own config. (cgroup has three values, but user has two).

Reply:
The original implementation had a single 300-odd like monolithic struct, but I tried to split it a bit to make things easier to parse. I don't know how successful this was.

Reply:
I love the general structure! the nits are more about how to split it up. (it's pretty much the same notes I had in https://github.com/containers/libpod/pull/4265